pkg/util/kubernetes/apiserver/apiserver.go

```diff
@@ -674,3 +690,181 @@ func (c *APIClient) GetARandomNodeName(ctx context.Context) (string, error) {
+func GetDeploy(cl *APIClient, name string, namespace string) ([]byte, error) {
+	var b bytes.Buffer
+
+	deploy, err := cl.Cl.AppsV1().Deployments(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		log.Debugf("Can't retrieve Deployment %v from the API server: %s", name, err.Error())
+		return nil, err
+	}
+
+	deployjson, err := json.Marshal(deploy)
+	deployyaml, err := yaml.JSONToYAML(deployjson)
+	fmt.Fprintln(&b, string(deployyaml))
+
+	return b.Bytes(), nil
+}
+
+// decodeRelease decodes the bytes of data into a readable byte array.
+// Data must contain a base64 encoded gzipped string of a valid release, otherwise nil is returned.
+func decodeRelease(data string) []byte {
+	// base64 decode string
+	b, err := b64.DecodeString(data)
+	if err != nil {
+		return nil
+	}
+
+	// For backwards compatibility with releases that were stored before
+	// compression was introduced we skip decompression if the
+	// gzip magic header is not found
+	if bytes.Equal(b[0:3], magicGzip) {
+		r, err := gzip.NewReader(bytes.NewReader(b))
+		if err != nil {
+			return nil
+		}
+		defer r.Close()
+		b2, err := io.ReadAll(r)
+		if err != nil {
+			return nil
+		}
+		b = b2
+	}
+	return b
+}
+
+// getConfigmapList retrieves a list of configmaps for a given namespace
+func getConfigmapList(cl *APIClient, namespace string) ([]v1.ConfigMap, error) {
+	configmaps, err := cl.Cl.CoreV1().ConfigMaps(namespace).List(context.TODO(), metav1.ListOptions{TimeoutSeconds: &cl.timeoutSeconds})
+	if err != nil {
+		log.Errorf("Can't list configmaps from the API server: %s", err.Error())
+		return nil, err
+	}
+	return configmaps.Items, nil
+}
+
+// getSecretList retrieves a list of secrets for a given namespace
+func getSecretList(cl *APIClient, namespace string) ([]v1.Secret, error) {
+	secrets, err := cl.Cl.CoreV1().Secrets(namespace).List(context.TODO(), metav1.ListOptions{TimeoutSeconds: &cl.timeoutSeconds})
+	if err != nil {
+		log.Errorf("Can't list secrets from the API server: %s", err.Error())
+		return nil, err
+	}
+	return secrets.Items, nil
+}
+
+// Retrieve Helm chart user values from the API server for a given name and namespace looking through Configmaps
+func GetDatadogHelmUserValuesFromConfigmap(cl *APIClient, name string, namespace string) ([]byte, error) {
+	var b bytes.Buffer
+	var userConfig chartUserValues
+	var selector labels.Selector
+	var helmConfigmap *v1.ConfigMap
+
+	// Only a single release for a given name can be deployed at one time.
+	// Other releases present a different status.
+	selector = labels.Set{
+		"owner":  "helm",
+		"status": "deployed",
+		"name":   name,
+	}.AsSelector()
+
+	configmaps, err := getConfigmapList(cl, namespace)
+	if err != nil {
+		log.Errorf("Can't list configmaps from the API server: %s", err.Error())
+	}
+
+	for _, cm := range configmaps {
+		if selector.Matches(labels.Set(cm.ObjectMeta.Labels)) {
```

Comment:
the selector can be already done in the apiserver query, so you should have only one configmap returned by the api server.

Reply:
you can check how it is done in the helm check https://github.com/DataDog/datadog-agent/blob/19bec7dd40d087109598094f9720f286ea9ea691/pkg/collector/corechecks/cluster/helm/helm.go#L206-L208

basically in the `getConfigmapList` function you can provide a labelselector/fieldselector to the "list" function

Reply:
👍 modified the query to avoid looping and instead directly get the sole configmap/secret